File: src/backend/app/schemas/experiment.py
"""
Pydantic schema models for experiment data validation, serialization, and API responses
in the Molecular Data Management and CRO Integration Platform.

This module defines the data models used for experiment creation, management, and submission
to CROs, with appropriate validation rules, type annotations, and documentation.
"""

from typing import Optional, List, Dict, Any, Union  # standard library
from datetime import datetime  # standard library
from uuid import UUID  # standard library

from pydantic import BaseModel, Field, validator  # pydantic version 2.0+

from ..constants import ExperimentStatus  # Internal import for status enumeration
from .experiment_type import ExperimentTypeRead  # Internal import for experiment type reference
from .user import UserResponse  # Internal import for user reference
from .molecule import MoleculeResponse  # Internal import for molecule reference


@validator('status', pre=True)
def validate_status(cls, v):
    """
    Validates that an experiment status value is one of the allowed values.
    
    Args:
        cls: The class being validated
        v: The value to validate
        
    Returns:
        The validated status if valid
        
    Raises:
        ValueError: If the status is not a valid ExperimentStatus value
    """
    if v is None:
        return None
    
    try:
        # Convert to uppercase for case-insensitive comparison
        v_upper = v.upper() if isinstance(v, str) else v
        # Check if it's a valid enum value
        valid_statuses = [status.name for status in ExperimentStatus]
        if v_upper not in valid_statuses:
            raise ValueError(f"Invalid status: {v}. Must be one of {valid_statuses}")
        return v_upper
    except (AttributeError, ValueError) as e:
        raise ValueError(f"Invalid status: {v}. Must be one of {[status.name for status in ExperimentStatus]}")


@validator('parameters', pre=True)
def validate_parameters(cls, v):
    """
    Validates that experiment parameters are properly formatted.
    
    Args:
        cls: The class being validated
        v: The value to validate (list of parameter dictionaries)
        
    Returns:
        The validated parameters list if valid
        
    Raises:
        ValueError: If parameters are not properly formatted
    """
    if v is None:
        return []
    
    if not isinstance(v, list):
        raise ValueError("Parameters must be a list")
    
    for param in v:
        if not isinstance(param, dict):
            raise ValueError("Each parameter must be a dictionary")
        
        if 'parameter_name' not in param or not param['parameter_name']:
            raise ValueError("Each parameter must have a non-empty 'parameter_name'")
        
        if 'parameter_value' not in param:
            raise ValueError("Each parameter must have a 'parameter_value'")
    
    return v


class ExperimentParameterBase(BaseModel):
    """
    Base Pydantic model for experiment parameter data with common fields.
    
    This model defines the basic structure for experiment parameters, including
    the parameter name and value fields.
    """
    parameter_name: str = Field(..., description="Name of the parameter")
    parameter_value: str = Field(..., description="Value of the parameter")


class ExperimentParameterCreate(ExperimentParameterBase):
    """
    Pydantic model for creating a new experiment parameter.
    
    Inherits all fields from ExperimentParameterBase.
    """
    pass


class ExperimentParameterResponse(ExperimentParameterBase):
    """
    Pydantic model for experiment parameter data in API responses.
    
    Extends the base model with additional fields needed for API responses.
    """
    id: UUID = Field(..., description="Unique identifier for the parameter")
    experiment_id: UUID = Field(..., description="ID of the experiment this parameter belongs to")


class ExperimentBase(BaseModel):
    """
    Base Pydantic model for experiment data with common fields.
    
    This model defines the basic structure for experiments, including the name,
    type, description, status, and parameters fields.
    """
    name: str = Field(..., description="Name of the experiment")
    type_id: UUID = Field(..., description="ID of the experiment type")
    description: Optional[str] = Field(None, description="Detailed description of the experiment")
    status: Optional[str] = Field(ExperimentStatus.DRAFT.name, description="Current status of the experiment")
    parameters: Optional[List[ExperimentParameterCreate]] = Field(default_factory=list, 
                                                               description="List of experiment parameters")


class ExperimentCreate(ExperimentBase):
    """
    Pydantic model for creating a new experiment.
    
    Extends the base model with fields needed for experiment creation, such as
    the creator ID and associated molecule IDs.
    """
    created_by: Optional[UUID] = Field(None, description="ID of the user creating the experiment")
    molecule_ids: Optional[List[UUID]] = Field(default_factory=list, 
                                           description="IDs of molecules to associate with the experiment")


class ExperimentUpdate(BaseModel):
    """
    Pydantic model for updating an existing experiment.
    
    All fields are optional to allow partial updates. Only provided fields
    will be updated in the database.
    """
    name: Optional[str] = Field(None, description="Updated name of the experiment")
    type_id: Optional[UUID] = Field(None, description="Updated experiment type ID")
    description: Optional[str] = Field(None, description="Updated description of the experiment")
    status: Optional[str] = Field(None, description="Updated status of the experiment")
    parameters: Optional[List[ExperimentParameterCreate]] = Field(None, 
                                                               description="Updated list of experiment parameters")
    molecule_ids: Optional[List[UUID]] = Field(None, 
                                           description="Updated list of molecule IDs to associate with the experiment")


class ExperimentInDBBase(ExperimentBase):
    """
    Base Pydantic model for experiment data as stored in the database.
    
    Extends the base model with additional fields that are present in the database,
    such as ID, creator ID, and timestamps.
    """
    id: UUID = Field(..., description="Unique identifier for the experiment")
    created_by: UUID = Field(..., description="ID of the user who created this experiment")
    created_at: datetime = Field(..., description="Timestamp when the experiment was created")
    updated_at: Optional[datetime] = Field(None, description="Timestamp when the experiment was last updated")

    class Config:
        orm_mode = True


class ExperimentResponse(ExperimentInDBBase):
    """
    Pydantic model for experiment data in API responses.
    
    Extends the database model with additional fields needed for API responses,
    such as the experiment type, parameters, creator details, and molecule count.
    """
    experiment_type: ExperimentTypeRead = Field(..., description="Details of the experiment type")
    parameters: List[ExperimentParameterResponse] = Field(default_factory=list, 
                                                      description="List of experiment parameters")
    creator: Optional[UserResponse] = Field(None, description="User who created this experiment")
    molecule_count: int = Field(..., description="Number of molecules associated with this experiment")

    class Config:
        orm_mode = True


class ExperimentDetailResponse(ExperimentInDBBase):
    """
    Pydantic model for detailed experiment data in API responses.
    
    Extends the database model with more comprehensive details, including
    the full list of associated molecules and submission history.
    """
    experiment_type: ExperimentTypeRead = Field(..., description="Details of the experiment type")
    parameters: List[ExperimentParameterResponse] = Field(default_factory=list, 
                                                      description="List of experiment parameters")
    creator: Optional[UserResponse] = Field(None, description="User who created this experiment")
    molecules: List[MoleculeResponse] = Field(default_factory=list, 
                                          description="List of molecules associated with this experiment")
    submissions: Optional[List[Dict]] = Field(default_factory=list, 
                                           description="List of submissions for this experiment")

    class Config:
        orm_mode = True


class ExperimentListResponse(BaseModel):
    """
    Pydantic model for paginated list of experiments in API responses.
    
    Used for paginated responses containing multiple experiments.
    Includes the total count, page number, and page size for pagination.
    """
    items: List[ExperimentResponse] = Field(..., description="List of experiments")
    total: int = Field(..., description="Total number of experiments matching the criteria")
    page: int = Field(..., description="Current page number")
    size: int = Field(..., description="Number of items per page")


class ExperimentFilter(BaseModel):
    """
    Pydantic model for filtering experiments by various criteria.
    
    This model defines the available filter parameters that can be used
    when querying experiments. All fields are optional to allow flexible filtering.
    """
    name: Optional[str] = Field(None, description="Filter by experiment name (partial match)")
    type_id: Optional[UUID] = Field(None, description="Filter by experiment type ID")
    status: Optional[str] = Field(None, description="Filter by experiment status")
    created_by: Optional[UUID] = Field(None, description="Filter by creator user ID")
    created_after: Optional[datetime] = Field(None, description="Filter by creation date (after)")
    created_before: Optional[datetime] = Field(None, description="Filter by creation date (before)")
    molecule_id: Optional[UUID] = Field(None, description="Filter by associated molecule ID")
    sort_by: Optional[str] = Field(None, description="Field to sort by")
    sort_desc: Optional[bool] = Field(False, description="Sort in descending order if true")

File: src/web/src/types/experiment.ts
/**
 * Experiment-related TypeScript definitions for the Molecular Data Management and CRO Integration Platform.
 * These types represent experiments, their parameters, statuses, and relationships with molecules and CROs.
 * @packageDocumentation
 */

import { Molecule } from './molecule';
import { User } from './user';

/**
 * Enumeration of experiment status values.
 * Represents the possible states of an experiment throughout its lifecycle.
 */
export enum ExperimentStatus {
  /** Initial state when experiment is being defined */
  DRAFT = 'draft',
  /** Experiment is queued for submission to CRO */
  QUEUED = 'queued',
  /** Experiment has been submitted to CRO */
  SUBMITTED = 'submitted',
  /** Experiment was rejected by CRO */
  REJECTED = 'rejected',
  /** CRO has provided a quote, awaiting approval */
  QUOTE_PENDING = 'quote_pending',
  /** Quote was rejected by the pharma user */
  QUOTE_REJECTED = 'quote_rejected',
  /** Experiment is being processed by CRO */
  IN_PROGRESS = 'in_progress',
  /** Experiment is complete, results in preparation */
  RESULTS_PENDING = 'results_pending',
  /** Results have been uploaded and are available */
  RESULTS_AVAILABLE = 'results_available',
  /** Results were rejected and need revision */
  RESULTS_REJECTED = 'results_rejected',
  /** Experiment is fully completed and reviewed */
  COMPLETED = 'completed',
  /** Experiment was cancelled */
  CANCELLED = 'cancelled'
}

/**
 * Interface for experiment type definition.
 * Represents the categories and types of experiments available in the system.
 */
export interface ExperimentType {
  /** Unique identifier for the experiment type */
  id: string;
  /** Name of the experiment type */
  name: string;
  /** Detailed description of the experiment type */
  description: string;
  /** Category grouping for the experiment type */
  category: string;
}

/**
 * Interface for experiment parameters.
 * Represents configurable parameters for experiments.
 */
export interface ExperimentParameter {
  /** ID of the experiment this parameter belongs to */
  experiment_id: string;
  /** Name of the parameter */
  parameter_name: string;
  /** Value of the parameter */
  parameter_value: string | number;
}

/**
 * Interface for the relationship between experiments and molecules.
 * Represents molecules included in an experiment.
 */
export interface ExperimentMolecule {
  /** ID of the experiment */
  experiment_id: string;
  /** ID of the molecule */
  molecule_id: string;
  /** Timestamp when the molecule was added to the experiment */
  added_at: string;
  /** Optional molecule data for expanded information */
  molecule?: Molecule;
}

/**
 * Interface for experiment data with basic information.
 * Core data structure representing an experiment in the system.
 */
export interface Experiment {
  /** Unique identifier for the experiment */
  id: string;
  /** Name of the experiment */
  name: string;
  /** ID of the experiment type */
  type_id: string;
  /** Current status of the experiment */
  status: ExperimentStatus;
  /** ID of the user who created the experiment */
  created_by: number;
  /** Timestamp when the experiment was created */
  created_at: string;
  /** Timestamp when the experiment was last updated */
  updated_at?: string;
  /** Experiment type information */
  type?: ExperimentType;
  /** Creator user information */
  creator?: User;
  /** Parameters for this experiment */
  parameters?: ExperimentParameter[];
  /** Count of molecules in this experiment */
  molecule_count?: number;
}

/**
 * Interface for experiment data with detailed information.
 * Provides comprehensive information about an experiment and its relationships.
 */
export interface ExperimentDetail extends Experiment {
  /** Molecules included in this experiment */
  molecules: ExperimentMolecule[];
  /** Submissions related to this experiment */
  submissions: any[]; // This would be replaced with a Submission interface
}

/**
 * Interface for creating a new experiment.
 * Contains the required fields to create an experiment.
 */
export interface ExperimentCreate {
  /** Name of the experiment */
  name: string;
  /** ID of the experiment type */
  type_id: string;
  /** Parameters for the experiment */
  parameters: Partial<ExperimentParameter>[];
  /** IDs of molecules to include in the experiment */
  molecule_ids: string[];
}

/**
 * Interface for updating an existing experiment.
 * Contains fields that can be modified after creation.
 */
export interface ExperimentUpdate {
  /** Updated name of the experiment */
  name?: string;
  /** Updated status of the experiment */
  status?: ExperimentStatus;
  /** Updated parameters for the experiment */
  parameters?: Partial<ExperimentParameter>[];
}

/**
 * Interface for adding molecules to an experiment.
 * Used in API requests to add molecules to an experiment.
 */
export interface ExperimentMoleculeOperation {
  /** IDs of molecules to add to the experiment */
  molecule_ids: string[];
}

/**
 * Interface for filtering experiments by various criteria.
 * Used in API requests to filter experiments.
 */
export interface ExperimentFilter {
  /** Filter by experiment name (partial match) */
  name?: string;
  /** Filter by experiment type ID */
  type_id?: string;
  /** Filter by experiment status */
  status?: ExperimentStatus;
  /** Filter by creator user ID */
  created_by?: number;
  /** Filter by creation date (after this date) */
  created_after?: string;
  /** Filter by creation date (before this date) */
  created_before?: string;
  /** Filter by experiments containing a specific molecule */
  contains_molecule_id?: string;
  /** Page number for pagination */
  page?: number;
  /** Number of items per page */
  page_size?: number;
  /** Field to sort by */
  sort_by?: string;
  /** Sort in descending order if true */
  sort_desc?: boolean;
}

/**
 * Interface for paginated list of experiments in API responses.
 * Used for returning large sets of experiment data with pagination.
 */
export interface ExperimentListResponse {
  /** Array of experiment items */
  items: Experiment[];
  /** Total number of experiments matching the query */
  total: number;
  /** Current page number */
  page: number;
  /** Number of items per page */
  size: number;
  /** Applied filters */
  filters: ExperimentFilter;
}

/**
 * Interface for list of experiment types in API responses.
 * Used for returning available experiment types.
 */
export interface ExperimentTypeListResponse {
  /** Array of experiment type items */
  items: ExperimentType[];
  /** Total number of experiment types */
  total: number;
}

/**
 * Interface for updating the status of an experiment.
 * Used in API requests to update experiment status.
 */
export interface ExperimentStatusUpdate {
  /** New status for the experiment */
  status: ExperimentStatus;
  /** Optional notes about the status change */
  notes?: string;
}

/**
 * Interface for experiment state in Redux store.
 * Represents the complete state of the experiment module in the application.
 */
export interface ExperimentState {
  /** Array of experiments */
  experiments: Experiment[];
  /** Array of experiment types */
  experimentTypes: ExperimentType[];
  /** Currently selected experiment */
  currentExperiment: ExperimentDetail | null;
  /** Loading state flag */
  loading: boolean;
  /** Error message if any */
  error: string | null;
  /** Total number of experiments */
  totalExperiments: number;
  /** Current page number */
  currentPage: number;
  /** Number of items per page */
  pageSize: number;
  /** Current filter criteria */
  filter: ExperimentFilter;
}